### PR TITLE
Update rest api endpoints ChainDataProvider and ValidatorDataProvider

### LINKED ISSUE #5606: Update rest api endpoints using ChainDataProvider and ValidatorDataProvider
All REST endpoints that use ChainDataProvider can potentially return a 204/503 error via the generic handler, when a `ChainDataUnavailableException` is raised, or a `ServiceUnavailableException`.
Any rest api endpoints that are affected will extend `MigratingEndpointAdapter`.

For all endpoints that use `ChainDataProvider` or `ValidatorDataProvider`, in their metadataBuilder, they should be calling `.withChainDataResponses()`, and should not be listing their own response of 204 (SC_NO_CONTENT) or 503 (SC_SERVICE_UNAVAILABLE).
The rest api handlers are located in `data/beaconrestapi/src/main/tech/pegasys/teku/handlers`. They'll all have private providers that look like one of:
`ValidatorDataProvider provider;`
`ChainDataProvider provider;`

The flow on effect is that the `MigratedOpenApiIntegrationTest` will require that files get updated to include the responses, basically the change is in output and just needs to be applied to the write resource paths for the test.

Unit tests should be added for the adapters that didn't list these return codes, they'll be basically:
```
  @Test
  void metadata_shouldHandle204() {
    verifyMetadataEmptyResponse(handler, SC_NO_CONTENT);
  }
  @Test
  void metadata_shouldHandle503() throws JsonProcessingException {
    verifyMetadataErrorResponse(handler, SC_SERVICE_UNAVAILABLE);
  }
```

### PR DESCRIPTION
**Pull request by KatyaRyazantseva**
## PR Description
All endpoints that use `ChainDataProvider` or `ValidatorDataProvider`, in their metadataBuilder are calling .withChainDataResponses() and are not listing their own response of 204 (SC_NO_CONTENT) or 503 (SC_SERVICE_UNAVAILABLE). 
Unit tests for these endpoints are added as well.

## Fixed Issue
fixes #5606

### CODE REVIEW

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostBlindedBlock.java [hunk 1]
```diff
@@ -59,6 +59,7 @@ public PostBlindedBlock(
   }
 
   @Override
+  // should be deleted ?
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostBlindedBlock.java]: handleRequest can stay, that bit services the actual request...

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostBlindedBlock.java [hunk 2]
```diff
@@ -114,8 +115,9 @@ broadcast but a different status code is returned (202). Pre-Bellatrix, this end
             SC_ACCEPTED,
             "Block has been successfully broadcast, but failed validation and has not been imported.")
         .withBadRequestResponse(Optional.of("Unable to parse request body."))
-        .response(
-            SC_SERVICE_UNAVAILABLE, "Beacon node is currently syncing.", HTTP_ERROR_RESPONSE_TYPE)
+        // .response(
+        //     SC_SERVICE_UNAVAILABLE, "Beacon node is currently syncing.", HTTP_ERROR_RESPONSE_TYPE)
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostBlindedBlock.java]: Yep we can remove the commented out bits, this is what the intent is for this change.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostBlock.java [hunk 3]
```diff
@@ -60,6 +60,7 @@ public PostBlock(
   }
 
   @Override
+  // should be deleted ?
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostBlock.java]: handleRequest will stay

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostBlock.java [hunk 4]
```diff
@@ -114,8 +115,9 @@ validation, a separate success response code (202) is used to indicate that the
             SC_ACCEPTED,
             "Block has been successfully broadcast, but failed validation and has not been imported.")
         .withBadRequestResponse(Optional.of("Unable to parse request body."))
-        .response(
-            SC_SERVICE_UNAVAILABLE, "Beacon node is currently syncing.", HTTP_ERROR_RESPONSE_TYPE)
+        // .response(
+        //     SC_SERVICE_UNAVAILABLE, "Beacon node is currently syncing.", HTTP_ERROR_RESPONSE_TYPE)
+        .withChainDataResponses()
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostBlock.java]: this is good, we can cleanup comment.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/validator/GetProposerDuties.java [hunk 5]
```diff
@@ -16,8 +16,10 @@
 import static tech.pegasys.teku.beaconrestapi.BeaconRestApiTypes.EPOCH_PARAMETER;
 import static tech.pegasys.teku.ethereum.json.types.validator.ProposerDutiesBuilder.PROPOSER_DUTIES_TYPE;
 import static tech.pegasys.teku.infrastructure.http.HttpStatusCodes.SC_OK;
+// should be deleted?
 import static tech.pegasys.teku.infrastructure.http.HttpStatusCodes.SC_SERVICE_UNAVAILABLE;
 import static tech.pegasys.teku.infrastructure.http.RestApiConstants.SERVICE_UNAVAILABLE;
+//
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/validator/GetProposerDuties.java]: when you run `./gradlew spotlessJavaApply` it'll remove unused import statements for you, so I'd suggest running that and a lot of these will be updated.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/validator/PostAttesterDuties.java [hunk 6]
```diff
@@ -78,6 +79,7 @@ public PostAttesterDuties(final DataProvider dataProvider) {
   }
 
   @Override
+  // should be deleted?
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/validator/PostAttesterDuties.java]: as above, keeping handleRequest fn

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/AbstractPostBlockV2.java [hunk 7]
```diff
@@ -41,6 +41,7 @@ public AbstractPostBlockV2(
 
   protected AsyncApiResponse processSendSignedBlockResult(final SendSignedBlockResult result) {
     return result
+        .withChainDataResponses() // ? no metadata
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/AbstractPostBlockV2.java]: This was an understandable guess, but not quite :)

Basically all of the implementations extending `AbstractPostBlock` define the metadata you were looking for, so this is not required here specifically.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlindedBlockV2.java [hunk 8]
```diff
@@ -61,6 +61,7 @@ public PostBlindedBlockV2(
   }
 
   @Override
+  // should be deleted ?
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlindedBlockV2.java]: keeping `handleRequest`

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlindedBlockV2.java [hunk 9]
```diff
@@ -125,8 +126,9 @@ broadcast but a different status code is returned (202). Pre-Bellatrix, this end
             SC_ACCEPTED,
             "Block has been successfully broadcast, but failed validation and has not been imported.")
         .withBadRequestResponse(Optional.of("Unable to parse request body."))
-        .response(
-            SC_SERVICE_UNAVAILABLE, "Beacon node is currently syncing.", HTTP_ERROR_RESPONSE_TYPE)
+        // .response(
+        //     SC_SERVICE_UNAVAILABLE, "Beacon node is currently syncing.", HTTP_ERROR_RESPONSE_TYPE)
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlindedBlockV2.java]: remove comment, it'll be in code history if we care to refer back, but its redundant.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlockV2.java [hunk 10]
```diff
@@ -62,6 +62,7 @@ public PostBlockV2(
   }
 
   @Override
+  // should be deleted ?
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlockV2.java]: same - keeping

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlockV2.java [hunk 11]
```diff
@@ -125,8 +126,9 @@ validation, a separate success response code (202) is used to indicate that the
             SC_ACCEPTED,
             "Block has been successfully broadcast, but failed validation and has not been imported.")
         .withBadRequestResponse(Optional.of("Unable to parse request body."))
-        .response(
-            SC_SERVICE_UNAVAILABLE, "Beacon node is currently syncing.", HTTP_ERROR_RESPONSE_TYPE)
+        // .response(
+        //     SC_SERVICE_UNAVAILABLE, "Beacon node is currently syncing.", HTTP_ERROR_RESPONSE_TYPE)
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlockV2.java]: cleanup  comment  block

**Review by rolfyone — COMMENTED** [review]
Thanks for having a look at this, we're off to a really good start! Most of the feedback is basically repeated so it's not nearly as bad as it seems (the 11 comment count)

The build process should pass spotless once `./gradlew spotlessJavaApply` is run, and then it'll cleanup all the unused imports, so it's worth running that prior to committing...

The tests look like what I was expecting, I haven't really looked super closely beyond spotless and the questions asked... Definitely on the right track here.

**Comment by rolfyone** [comment]
you'll also find when you run `./gradlew integrationTest` that it'll break some of `OpenApiIntegrationTest` (iirc) and it's basically a set of reference files that will now show you having extra responses that aren't currently in some reference files - so you basically are going to need to update those reference json files (less bad than it sounds - the json diff is hopefully fairly instructive)

File: data/beaconrestapi/src/integration-test/resources/tech/pegasys/teku/beaconrestapi/beacon/paths/_eth_v1_beacon_blinded_blocks.json [hunk 12]
```diff
@@ -1,92 +1,115 @@
...
         }
       },
-      "503" : {
-        "description" : "Beacon node is currently syncing.",
-        "content" : {
-          "application/json" : {
-            "schema" : {
-              "$ref" : "#/components/schemas/HttpErrorResponse"
+      "503": {
+        "description": "Service unavailable",
```
Comment (KatyaRyazantseva) [inline data/beaconrestapi/src/integration-test/resources/tech/pegasys/teku/beaconrestapi/beacon/paths/_eth_v1_beacon_blinded_blocks.json]: @rolfyone drawing your attention to 503 response. For APIs with 503 being implemented the description has changed.

File: data/beaconrestapi/src/integration-test/resources/tech/pegasys/teku/beaconrestapi/beacon/paths/_eth_v1_beacon_rewards_attestations_{epoch}.json [hunk 13]
```diff
@@ -1,85 +1,94 @@
...
-              "type": "string"
+    } ],
+    "requestBody" : {
+      "required" : false,
+      "content" : {
+        "application/json" : {
+          "schema" : {
+            "type" : "array",
+            "items" : {
+              "type" : "string"
```
Comment (rolfyone) [inline data/beaconrestapi/src/integration-test/resources/tech/pegasys/teku/beaconrestapi/beacon/paths/_eth_v1_beacon_rewards_attestations_{epoch}.json:25]: this one still has that formatting from manually doing things i think...

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/node/GetHealth.java [hunk 14]
```diff
@@ -65,7 +72,7 @@ public GetHealth(final DataProvider provider) {
   public void handleRequest(final RestApiRequest request) throws JsonProcessingException {
     request.header(Header.CACHE_CONTROL, CACHE_NONE);
     if (!chainDataProvider.isStoreAvailable() || syncProvider.getRejectedExecutionCount() > 0) {
-      request.respondWithCode(SC_SERVICE_UNAVAILABLE);
+      request.respondError(SC_SERVICE_UNAVAILABLE, "Node not initialized or having issues");
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/node/GetHealth.java]: nit: Im not sure we need to change this.... not having the error message is fine I think for service_unavailable... 
I'd probably suggest just putting the original 503 response line back in and not having the error here...

File: data/beaconrestapi/src/integration-test/resources/tech/pegasys/teku/beaconrestapi/beacon/paths/_eth_v1_beacon_rewards_attestations_{epoch}.json (hunk 13 above)
Comment (KatyaRyazantseva) [reply data/beaconrestapi/src/integration-test/resources/tech/pegasys/teku/beaconrestapi/beacon/paths/_eth_v1_beacon_rewards_attestations_{epoch}.json:25]: double checked, indeed, the red part seems to be edited manually. The green one is the same as after "but was..."

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/node/GetHealth.java (hunk 14 above)
Comment (KatyaRyazantseva) [reply data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/node/GetHealth.java]: Ok, will roll back GetHealth and Readiness.

**Comment by rolfyone** [comment]
I think we're up to just resolving that conflict, happy to do that?

**Comment by KatyaRyazantseva** [comment]
> I think we're up to just resolving that conflict, happy to do that?

I've finished, struggling with spotlessJavaApply. Will push the updates soon.

**Comment by rolfyone** [comment]
> > I think we're up to just resolving that conflict, happy to do that?
> 
> I've finished, struggling with spotlessJavaApply. Will push the updates soon.

can you try spotless with your code editor closed and see if that makes a difference?

**Comment by KatyaRyazantseva** [comment]
> can you try spotless with your code editor closed and see if that makes a difference?

I can, but I haven't caught the logic yet. It can delete a file unexpectedly. When I bring it back, it doesn't delete it anymore.

File: data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlockV2Test.java [hunk 15]
```diff
@@ -101,4 +106,14 @@
 
     assertThat(request.getResponseCode()).isEqualTo(SC_BAD_REQUEST);
   }
+
+  @Test
+  void metadata_shouldHandle204() {
```
[the remark this answers is not in the document]
Comment (rolfyone) [reply data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlockV2Test.java]: fixed.

File: data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlockV2Test.java [hunk 16]
```diff
@@ -101,4 +106,14 @@
...
+    verifyMetadataEmptyResponse(handler, SC_NO_CONTENT);
+  }
+
+  @Test
+  void metadata_shouldHandle503() throws JsonProcessingException {
```
[the remark this answers is not in the document]
Comment (rolfyone) [reply data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/PostBlockV2Test.java]: fixed

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/node/GetHealth.java [hunk 17]
```diff
@@ -55,6 +56,8 @@ public GetHealth(final DataProvider provider) {
             .response(SC_OK, "Node is ready")
             .response(SC_PARTIAL_CONTENT, "Node is syncing but can serve incomplete data")
             .response(SC_SERVICE_UNAVAILABLE, "Node not initialized or having issues")
+            .response(
+                SC_NO_CONTENT, "Data is unavailable because the chain has not yet reached genesis")
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/node/GetHealth.java:60]: not sure about this one..

Comment (rolfyone) [reply data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/node/GetHealth.java:60]: oh nvm i get it...

**Review by rolfyone — APPROVED** [review]
LGTM

**Pull request merged**